### ISSUE TITLE
Unable to find/read data

### ISSUE THREAD

**Issue by vkuznet** [scikit-hep/uproot#15]
Hi, I'm trying to read CMSSW GEN-SIM-RECO ROOT file (/afs/cern.ch/user/v/valya/public/E6B7DD7A-D398-E611-BD88-0025905B85DE.root)
with uproot and I know from ROOT object Browser that it contains data, e.g. it has recoTracks_generalTracks__RECO branch which has
two other branches recoTracks_generalTracks__RECO.present and recoTracks_generalTracks__RECO.obj, see https://www.dropbox.com/s/3xct7jskrelne56/RecoTracks.png?dl=0

Now, I read the file and dumped all branches, I can see the data for  recoTracks_generalTracks__RECO.present(it is array of booleans, all Trues) but  recoTracks_generalTracks__RECO.obj branch does not exists. Instead I see  recoTracks_generalTracks__RECO. branch. If I access it it contains empty array.

```
    with uproot.open(fname) as tree:
        eTree = tree['Events']
        names = ['recoTracks_generalTracks__RECO.present', 'recoTracks_generalTracks__RECO.', 'recoTracks_generalTracks__RECO.obj']
        for bName in names:
            print("Branch {} exists {}".format(bName, bName in eTree.branchnames))
            data = eTree.array(bName)
            print(data)
```
and I got the following:
```
Branch recoTracks_generalTracks__RECO.present exists False
[ True  True  True  True  True  True  True  True  True  True  True  True...]
Branch recoTracks_generalTracks__RECO. exists True
[]
Branch recoTracks_generalTracks__RECO.obj exists False
Traceback (most recent call last):
  File "./vk_test.py", line 41, in <module>
    data = eTree.array(bName)
  File "/Users/vk/Work/Languages/Python/GIT/uproot/uproot/tree.py", line 529, in array
    out, = self.arrays(branchdtypes=branchdtypes, executor=executor, outputtype=tuple, block=block)
  File "/Users/vk/Work/Languages/Python/GIT/uproot/uproot/tree.py", line 447, in arrays
    for branch, dtype in self._normalizeselection(branchdtypes, self.allbranches):
  File "/Users/vk/Work/Languages/Python/GIT/uproot/uproot/tree.py", line 266, in _normalizeselection
    raise ValueError("cannot produce an array from branch {0}".format(repr(name)))
ValueError: cannot produce an array from branch 'recoTracks_generalTracks__RECO.obj'
```

So, there are few questions:
- why recoTracks_generalTracks__RECO.present is not present in branchnames list, but it is listed if I iterate over tree arrays keys like `tree.arrays().keys()`
- why this branch actually contains data
- why uproot does not see Branch recoTracks_generalTracks__RECO.obj
- why it provides recoTracks_generalTracks__RECO. branch which is not shown in ROOT object Browser
- if recoTracks_generalTracks__RECO. is indeed recoTracks_generalTracks__RECO.obj why it is empty, since I can see data in ROOT object browser.

I understand traceback because I'm trying to access non-existing branch.

**Comment by jpivarski** [comment]
I'll take a look at that file soon, but maybe it would help if I start with a quick reply.

The uproot 1.x behavior was to attempt to assign Numpy dtypes to each branch when it is first loaded. Branch types that aren't numerical don't get a dtype attribute and are excluded from the "TTree.branches" list. This has the advantage that if you ask for all branches from a file where you can read 99% of the branches, you get them automatically, but it has the disadvantage that it's implicit and contrary to user expectations. Implicit is bad.

The uproot 2.0 behavior (see "file-interface" branch in GitHub; it's coming along!) is to not do any interpretation upon first reading, and in fact not have any permanent interpretation associated with the branch. Interpretations are customizable and a reasonable default is applied on the fly if not supplied. (An uproot "interpretation" object takes the place of Numpy "dtypes".) uproot 2.0 also reads TStreamerInfo, unlike uproot 1.x, so in principle every branch should get a reasonable default interpretation.

If you're reading a complex file like CMS MiniAOD, I think you'll want to wait for 2.0, and you'd be happier if I finish that (few weeks?), rather than provide a work-around for 1.x.

Again, I'll look more deeply into this file soon.

**Comment by jpivarski** [comment]
Oh, that Dropbox link is a picture of a ROOT window. I'll have to find a RECO file. I have a MiniAOD file and see branches like

```
patMuons_slimmedMuons__PAT.
patMuons_slimmedMuons__PAT.present
patMuons_slimmedMuons__PAT.obj
patMuons_slimmedMuons__PAT.obj.m_state.vertex_.fCoordinates.fX
patMuons_slimmedMuons__PAT.obj.m_state.vertex_.fCoordinates.fY
patMuons_slimmedMuons__PAT.obj.m_state.vertex_.fCoordinates.fZ
patMuons_slimmedMuons__PAT.obj.m_state.p4Polar_.fCoordinates.fPt
patMuons_slimmedMuons__PAT.obj.m_state.p4Polar_.fCoordinates.fEta
patMuons_slimmedMuons__PAT.obj.m_state.p4Polar_.fCoordinates.fPhi
patMuons_slimmedMuons__PAT.obj.m_state.p4Polar_.fCoordinates.fM
```

**Comment by vkuznet** [comment]
Jim I listed file in my issue, it is available on AFS

/afs/cern.ch/user/v/valya/public/E6B7DD7A-D398-E611-BD88-0025905B85DE.root

**Comment by jpivarski** [comment]
I'm examining your file on lxplus with uproots 1 and 2. I see all of these branches in the `.allbranches` list, so maybe that policy I described above changed with version. In the future, no branches will be hidden.

uproot-1 is currently more capable at reading out data, and it says that `recoTracks_generalTracks__RECO.` has `int32` type, `recoTracks_generalTracks__RECO.obj` is unknown, and `recoTracks_generalTracks__RECO.present` is `bool`. I guess `.present` is an existence mask (all `True` values here) and the `recoTracks_generalTracks__RECO.` is very likely the counter for split data in `.obj`. The trouble is, the data are not split. Maybe that's why the array is empty. (This is the first time I've worked with an unsplit file.)

uproot-2 can tell me about classes and streamers. It says that all three have `vector<reco::Track>` type.

```
>>> t["recoTracks_generalTracks__RECO.present"].fClassName
'edm::Wrapper<vector<reco::Track> >'
>>> t["recoTracks_generalTracks__RECO.obj"].fClassName
'edm::Wrapper<vector<reco::Track> >'
>>> t["recoTracks_generalTracks__RECO."].fClassName
'edm::Wrapper<vector<reco::Track> >'
```

It figures. If `.obj` were split, the parent branch would probably have counters, and in unsplit mode, it's a placeholder empty array. The `.obj` has no subbranches, as it does in the split files I've looked at, so it's non-numerical. That's why uproot-1 didn't give it a dtype.

uproot-2 has read the TStreamerInfo for `reco::Track` and `vector<reco::Track>`, but I still need to use that to generate code that would extract full objects from the unsplit branch. In pure Python, this **will not** be fast because it will be sequential Python statements, not a Numpy call. Numba will help, and to what degree I don't yet know. Also, I should probably be splitting it on the fly so that subsequent work on these data can take advantage of Numpy.

So if you want to read this file, you'll have to wait for uproot-2 and install Numba. I'm going to close this issue for now and alert you when uproot-2 is ready, and I'll take your file and `recoTracks_generalTracks__RECO.obj` as a test case to try to handle first. (Initial success when we can read out the five track parameters.)
